src/data/finvec.lean

```diff
@@ -0,0 +1,193 @@
+  (h : Π (x : finvec n α) (y : finvec m α), C (x ++ y)) (z : finvec (n + m) α) : C z :=
+left_append_right z ▸ h z.left z.right
+
+@[simp] lemma left_append {x : finvec n α} {y : finvec m α} : (x ++ y).left = x :=
+congr_arg prod.fst (append_equiv.apply_symm_apply (x, y) : _)
+
+@[simp] lemma right_append {x : finvec n α} {y : finvec m α} : (x ++ y).right = y :=
+congr_arg prod.snd (append_equiv.apply_symm_apply (x, y) : _)
+
+lemma prod_ext {z z' : finvec (n + m) α} :
+  z = z' ↔ z.left = z'.left ∧ z.right = z'.right :=
+append_equiv.apply_eq_iff_eq.symm.trans prod.mk.inj_iff
+
+lemma append.inj_iff {x x' : finvec n α} {y y' : finvec m α} :
+  x ++ y = x' ++ y' ↔ x = x' ∧ y = y' :=
+by simp only [append, equiv.apply_eq_iff_eq, prod.mk.inj_iff]
+
+lemma left_zero {x : finvec (n + 0) α} : x.left = x :=
+by { ext ⟨_, _⟩, refl }
+
+lemma right_zero {x : finvec (0 + n) α} : x.right = x.cast (zero_add n) :=
+funext $ λ _, congr_arg x (by rw fin.nat_add_zero)
+
+@[simp] lemma append_zero {x : finvec n α} {y : finvec 0 α} : x ++ y = x :=
+begin
+  rw [prod_ext, left_append, left_zero, right_append],
+  exact ⟨rfl, subsingleton.elim _ _⟩
+end
+
+end prod
+
+section snoc
+
+/-! ### Appending a single element (snoc)
+
+This section is parallel to the `prod` section but based on the isomorphism αⁿ⁺¹ ≅ αⁿ × α.
+We name the projections `init : finvec (n+1) α → finvec n α`
+(which equals `left` specialized to `m = 1`) and `last : finvec (n+1) α → α`,
+with constructor `snoc : finvec n α → α → finvec (n+1) α` and
+corresponding induction principle `snoc_induction`. -/
+
+variables {n : ℕ}
+
+/-- All but the last coordinate of a `finvec`. -/
+protected def init : finvec (n + 1) α → finvec n α :=
+λ x j, x (fin.cast_succ j)
+
+lemma left_eq_init {x : finvec (n + 1) α} : x.left = x.init := rfl
+
+/-- The last coordinate of a `finvec`. -/
+protected def last : finvec (n + 1) α → α :=
+λ x, x (fin.last n)
+
+/-- The canonical equivalence `αⁿ⁺¹ ≅ αⁿ × α`. -/
+def snoc_equiv : finvec (n + 1) α ≃ finvec n α × α :=
+calc  (fin (n + 1) → α)
+    ≃ (fin n → α) × (fin 1 → α)  : append_equiv
+... ≃ (fin n → α) × α            : equiv.prod_congr (equiv.refl _) (equiv.fun_unique (fin 1) α)
+
+/-- Form a `finvec` by appending a single coordinate to an existing `finvec`. -/
+def snoc (x : finvec n α) (a : α) : finvec (n + 1) α :=
+snoc_equiv.symm (x, a)
+
+lemma snoc_equiv_app {z : finvec (n + 1) α} : snoc_equiv z = (z.init, z.last) := rfl
+
+lemma snoc_equiv_symm_app {x : finvec n α} {a : α} :
+  snoc_equiv.symm (x, a) = x.snoc a :=
+rfl
+
+@[simp] lemma init_snoc_last (x : finvec (n + 1) α) : x.init.snoc x.last = x :=
+snoc_equiv.symm_apply_apply x
+
+/-- Induction principle for `finvec (n + 1) α`,
+imagined as a structure containing its initial part and last coordinate. -/
+@[elab_as_eliminator]
+protected lemma snoc_induction {C : finvec (n + 1) α → Prop}
+  (h : Π (x : finvec n α) (a : α), C (snoc x a)) (z : finvec (n + 1) α) : C z :=
+init_snoc_last z ▸ h z.init z.last
```

Comment:
Thanks, I tried this generalization originally but abandoned it because of the `rw` in the definition. Didn't realize that you can still write a nice propositional beta lemma for it.